subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java

```diff
@@ -86,35 +89,57 @@ public void run() {
     }
 
     @Override
-    public boolean isCacheEnabled() {
-        return !cacheIfSpec.getSpecs().isEmpty() && cacheIfSpec.isSatisfiedBy(task)
-            && (doNotCacheIfSpec.isEmpty() || !doNotCacheIfSpec.isSatisfiedBy(task));
-    }
+    public TaskOutputCaching getCaching() {
+        if (cacheIfSpecs.isEmpty()) {
+            return DefaultTaskOutputCaching.notCacheable("Caching has not been enabled for the task");
+        }
+        if (!hasDeclaredOutputs()) {
+            return DefaultTaskOutputCaching.notCacheable("No outputs declared");
+        }
 
-    @Override
-    public boolean isCacheAllowed() {
         for (TaskPropertySpec spec : getFileProperties()) {
             if (spec instanceof NonCacheableTaskOutputPropertySpec) {
-                return false;
+                return DefaultTaskOutputCaching.notCacheable("Declares multiple output files for a single output property via `@OutputFiles`, `@OutputDirectories` or `TaskOutputs.files()`");
+            }
+        }
+        for (SelfDescribingSpec<TaskInternal> selfDescribingSpec : cacheIfSpecs) {
+            if (!selfDescribingSpec.isSatisfiedBy(task)) {
+                return DefaultTaskOutputCaching.notCacheable(selfDescribingSpec.getDisplayName());
```

Comment:
One weird thing about this.  The descriptions for `doNotCacheIf` specs are the reason for why the task isn't cacheable.  It would seem that the descriptions for `cacheIf` specs should be the reason for why the task *is* cacheable, but we're using it for the opposite here.

Maybe with something like:

```
myTask {
   cacheIf("on CI") {
       System.getenv("CI") != null
   }
}
```

We could say the reason is "Not caching task ':myTask': cacheIf not satisfied 'on CI'".

I'm not sure this is any better than what you already have.  I think it's clear enough that the reason is the reason for disabling caching.